### PR TITLE
BUG: Fix csv.QUOTE_NONNUMERIC quoting in to_csv

### LINKED ISSUE #12922: DataFrame.to_csv(quoting=csv.QUOTE_NONNUMERIC) quotes numeric values
#### Failing test

```
def test_pandas():
    import tempfile
    import csv
    import pandas as pd
    import numpy as np

    df = pd.DataFrame.from_dict({'column': [1.0, 2.0]})
    assert df['column'].dtype == np.dtype('float')

    with tempfile.TemporaryFile() as f:
        df.to_csv(f, quoting=csv.QUOTE_NONNUMERIC, index=False)

        f.seek(0)
        lines = f.read().splitlines()
        assert lines[0] == '"column"'
        assert not lines[1].startswith('"') # <--- THIS FAILS
        assert [1, 2] == map(float, lines[1:])
```

The issue is that the floats are being output wrapped with quotes, even though I requested QUOTE_NONNUMERIC.

The problem is that `pandas.core.internals.FloatBlock.to_native_types` (and by extension `pandas.formats.format.FloatArrayFormatter.get_result_as_array`) unconditionally formats the float array to a str array, which is then passed unchanged to the `csv` module and hence will be wrapped in quotes by that code.

I'm not 100% sure but the fix may be to have `FloatBlock.to_native_types` check if quoting is set, and if so to skip using the `FloatArrayFormatter`? I say this because `pandas.indexes.base.Index._format_native_types` already has a special case along these lines. This does seem a bit dirty though!

Here is an awful monkeypatch that works around the problem:

```
orig_to_native_types = pd.core.internals.FloatBlock.to_native_types
def to_native_types(self, *args, **kwargs):
    if kwargs.get('quoting'):
        values = self.values
        slicer = kwargs.get('slicer')
        if slicer is not None:
            values = values[:, slicer]

        return values

    res = orig_to_native_types(self, *args, **kwargs)
    print 'FloatBlock.to_native_types', args, kwargs, '=', res
    return res
pd.core.internals.FloatBlock.to_native_types = to_native_types
```

#### output of ``pd.show_versions()``

```
commit: None
python: 2.7.9.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 42 Stepping 7, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.0
nose: None
pip: 8.1.1
setuptools: 7.0
Cython: 0.20.1
numpy: 1.11.0
scipy: 0.13.3
statsmodels: None
xarray: None
IPython: 3.2.1
sphinx: None
patsy: 0.3.0
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: 1.0.0
tables: None
numexpr: None
matplotlib: 1.3.1
openpyxl: 2.0.4
xlrd: 0.9.2
xlwt: None
xlsxwriter: None
lxml: 3.3.2
bs4: 4.2.0
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.11
pymysql: None
psycopg2: None
jinja2: 2.7.2
boto: None
```

### PR DESCRIPTION
**Pull request by gfyoung**
Closes #12922: "bug" traced to #12194 via bisection, where the float formatting was unconditionally casting everything to string.

I say "bug" (with quotations) because the changes to `get_result_as_array` did **correctly** cast everything to string as per the documentation (i.e. it had inadvertently patched a bug itself even though it was just a cleaning PR).  However, the changes had overlooked the impact it would have on `to_csv`.

### CODE REVIEW

File: pandas/formats/format.py [hunk 1]
```diff
@@ -2138,6 +2016,31 @@ def formatter(value):
 
         return formatter
 
+    def _get_formatted_data(self):
+        """
+        Returns formatted float values. This is different from
```
Comment (jreback) [inline pandas/formats/format.py]: why are you making this a separate function ? the point was to eliminate this kind of thing;

File: doc/source/whatsnew/v0.18.2.txt [hunk 2]
```diff
@@ -101,6 +101,7 @@ API changes
 ~~~~~~~~~~~
 
 
+- ``DataFrame.to_csv()`` has dropped the ``engine`` parameter (:issue:`13418`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: this is a prior deprecation removal

do in s separate pr

File: pandas/formats/format.py (hunk 1 above)
Comment (gfyoung) [inline pandas/formats/format.py]: I think my reasoning was made clear in the doc-string.  `get_result_as_array` returns only strings (it supposed to be doing this for quite some time).  However, formatting CSV output is **not always** string.

Comment (gfyoung) [inline pandas/formats/format.py]: The point you refer to was well-intended but ultimately incorrect, and sure, we could add this to `get_result_as_array`, but that's an API change then.

Comment (jreback) [inline pandas/formats/format.py]: how so? it's an internal routine

File: doc/source/whatsnew/v0.18.2.txt (hunk 2 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.2.txt]: It made it easier to see what I was doing when I was fixing the bug in `to_csv` though (didn't have to consider the other `engine`).

File: pandas/formats/format.py (hunk 1 above)
Comment (gfyoung) [inline pandas/formats/format.py]: `get_result_as_array` - I wouldn't know that from the signature (no preceding `_`)

File: doc/source/whatsnew/v0.18.2.txt (hunk 2 above)
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: I get it but should do that change for 0.19 (should have done it for 0.18 but that ship sailed)

File: pandas/formats/format.py (hunk 1 above)
Comment (jreback) [inline pandas/formats/format.py]: these entire module is internal - in reality none should be leading _

we have very inconsistent usage

File: doc/source/whatsnew/v0.18.2.txt (hunk 2 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.2.txt]: No API changes like this in a minor release?  It has been deprecated for a long time though...do changes like these only go for major releases?

File: pandas/formats/format.py (hunk 1 above)
Comment (gfyoung) [inline pandas/formats/format.py]: Ah, I see.  Fair enough, then a note should be made about that in the module doc-string.  I'll move this functionality then into `get_result_as_array` then.

File: doc/source/whatsnew/v0.18.2.txt (hunk 2 above)
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: it's a prior deprecation removal / so no

that would totally break if you had ignored the warning and not fixed your code

people should have done it - but can't hurt to put in major version

File: pandas/formats/format.py (hunk 1 above)
Comment (jreback) [inline pandas/formats/format.py]: that's fine

File: doc/source/whatsnew/v0.18.2.txt (hunk 2 above)
Comment (gfyoung) [inline doc/source/whatsnew/v0.18.2.txt]: True, though the ignoring warning and breaking all of my code argument could be made for a major release as well...but fair enough: will break this up.

File: pandas/formats/format.py (hunk 1 above)
Comment (gfyoung) [inline pandas/formats/format.py]: Added note to `format.py` - unfortunately, integrating the no-format option from my new function into `get_result_as_array` caused a whole slew of test failures.  So I just moved it into `to_native_types` instead.

**Comment by gfyoung** [comment]
@jreback : Moved `engine` parameter removal into separate PR and removed the `_get_formatted_data` function and move its code in `to_native_types`.  Travis is still happy, so ready to merge if there are no other concerns.

File: pandas/formats/format.py [hunk 3]
```diff
@@ -1,4 +1,8 @@
 # -*- coding: utf-8 -*-
+"""
+Internal module for formatting output data (e.g. `to_csv`).
+"""
```
Comment (jreback) [inline pandas/formats/format.py]: this is not a very helpful comment.

File: pandas/core/internals.py [hunk 4]
```diff
@@ -1529,6 +1529,18 @@ def to_native_types(self, slicer=None, na_rep='', float_format=None,
         if slicer is not None:
             values = values[:, slicer]
 
+        # no special formatting is desired
+        if float_format is None and decimal == '.':
```
Comment (jreback) [inline pandas/core/internals.py]: I guess quoting *only* matters in a csv (and not generally to output, which is where the Formatters matter). pls add a comment with the issue number

File: pandas/formats/format.py (hunk 3 above)
Comment (gfyoung) [inline pandas/formats/format.py]: What should I add exactly?

Comment (jreback) [inline pandas/formats/format.py]: you can remove the ``to_csv`` (or add that it provides formatters for csv, html, latex and well as display formatting). It doesn't support ``to_csv``, rather the implemenation for it.

File: pandas/core/internals.py (hunk 4 above)
Comment (gfyoung) [inline pandas/core/internals.py]: Yep, done.

File: pandas/formats/format.py (hunk 3 above)
Comment (gfyoung) [inline pandas/formats/format.py]: Sure thing.  Done.

**Comment by gfyoung** [comment]
@jreback : Modified the relevant doc-strings and comments as requested.  Ready to merge if there are no other concerns.

**Comment by gfyoung** [comment]
@jreback : Modified the relevant doc-strings and comments as requested.  Ready to merge if there are no other concerns.

File: pandas/tests/frame/test_to_csv.py [hunk 5]
```diff
@@ -1131,3 +1102,83 @@ def test_to_csv_with_dst_transitions(self):
...
+
+    def test_to_csv_quoting(self):
+        df = DataFrame({
+            'c_string': ['a', 'b,c'],
+            'c_int': [42, np.nan],
+            'c_float': [1.0, 3.2],
+            'c_bool': [True, False],
+        })
+
+        expected = """\
```
Comment (TomAugspurger) [inline pandas/tests/frame/test_to_csv.py]: Style nit, you can use

```
expected = textwrap.dedent(
    """\
    ,c_bool,c_float,c_int,c_string
    0,True,1.0,42.0,a
    1,False,3.2,,"b,c"
    """
)
```

to avoid the odd indentation. Not worth holding up a merge though, if you aren't able to change.

Comment (gfyoung) [inline pandas/tests/frame/test_to_csv.py]: Odd indentation is the MO it seems for anything CSV-related.  I'll leave for now unless there are really strong objections.  But thanks for mentioning though @TomAugspurger !

**Comment by jreback** [comment]
ty
